Added created and modified dates to metadata extractor.

ciferkey opened the pull request:
Implementing additional meta data fields for creation and modification dates.

Should close out https://github.com/chimbori/crux/issues/20. I needed this for myself as well.

Followed the details in CONTRIBUTING.md.

---
Specifically handling the following metadata:

- itemprop=dateCreated
- itemprop=dateModified 
- property=article:published_time 
- property=article:modified_time 
- property=article:published 
- property=article:modified

I couldn't find any documentation online detailing a full list of possible tags to look for. This was based on an investigation of several big news sites.  

Current design stores the dates as a string so the consumer can decide how to parse it. It looks like [dateCreated](https://schema.org/dateCreated) follows a specific timestamp format. OpenGragh [article:published_time](https://ogp.me/#datetime) seems to have a specific format as well. NBC however was a weird example. It uses "article:published" without "_time" and has a completely different timestamp format. Do you think it makes sense to consider date parsing out of scope for the project?

I added a new html to the integration tests and updated some existing tests to check for the dates. I can add more examples to existing tests if you would like.

Gradle check and test tasks are showing green for me locally. Anything else you would like to verify the changes?

ciferkey committed:
Added created and modified dates to metadata extractor.

ciferkey committed:
Added created and modified dates to metadata extractor.

chimbori commented on src/test/kotlin/com/chimbori/crux/articles/GoldenFilesTest.kt:
```diff
@@ -510,6 +514,22 @@ class GoldenFilesTest {
+        "The U.S. is stepping up its efforts to address the snowballing monkeypox outbreak — the country's largest to date by far, with 700 confirmed cases.",
+        document?.text()
+      )
+      assertEquals("Fri Jul 08 2022 00:44:04 GMT+0000 (UTC)", created)
```
It would be good to have all dates be ISO-8601 formatted, that way callers don’t have to wonder which way to parse a date, or know about multiple formats.

chimbori commented on test_data/nbc.html:
```diff
@@ -0,0 +1,5 @@
+<!DOCTYPE html><html lang="en"><head><link href="https://nodeassets.nbcnews.com/_next/static/css/38dbe5b2fda7a55d.css" rel="preload" as="style"/><style>@font-face{font-family:iconfont;src:url(https://nodeassets.nbcnews.com/_next/static/assets/fonts/iconfont.17f6dcc97d.woff2)format("woff2"),url(https://nodeassets.nbcnews.com/_next/static/assets/fonts/iconfont.5950e1148d.woff)format("woff");font-di ...
```
This particular example looks like a big chunk of JSON, though the date is in HTML <meta> tags. So apart from the meta tags, the rest of the doc isn’t parseable as HTML at all. Any chance we could use an example whose content can be parsed successfully as well? (I don’t believe it should be in scope for Crux to parse JSON-blob SPAs).

ciferkey replied:
> I don’t believe it should be in scope for Crux to parse JSON-blob SPAs

I did some more digging and it looks like that blob is for [JSON LB](https://json-ld.org/) rather than a blob for SPAs. Feels very much like [standards](https://xkcd.com/927/). It seems feasible to add support but definetely out of scope for this particular PR.

I dug through some more sites and couldn't find any others that use "article:published" rather than "article:published_at". I'd be OK with dropping parsing that and "article:modified". Since its not covered by a spec I can find I'm OK with dropping support for it and removing the NBC example. That will also simply date string parsing since the four tags we would support all need to be ISO 8601 by spec. 

What do you think?

chimbori replied:
That sounds like a pretty good simplification of this CL, plus it can easily be added later. Good idea to drop this!

Once you submit this PR, I will look into some basic validation for testing whether parsed dates are ISO 8601 or not (and convert if necessary).

chimbori commented:
This is awesome, thanks for the detailed write-up and of course the code!

> Current design stores the dates as a string so the consumer can decide how to parse it. It looks like [dateCreated](https://schema.org/dateCreated) follows a specific timestamp format. OpenGragh [article:published_time](https://ogp.me/#datetime) seems to have a specific format as well.

In fact, both of them have standardized on ISO 8601, which is a good design decision. As you noted, NBC is using a different format ([deprecated RFC 7231](https://httpwg.org/specs/rfc7231.html#http.date)) so it would be ideal to convert this to ISO 8601 before providing to callers.

chimbori commented on src/main/kotlin/com/chimbori/crux/api/Fields.kt:
```diff
@@ -8,6 +8,8 @@ public object Fields {
   public const val LANGUAGE: String = "language"
   public const val DISPLAY: String = "display"
   public const val ORIENTATION: String = "orientation"
+  public const val CREATED: String = "created"
```
How about calling this `PUBLISHED_AT` instead of `CREATED`, so it uses the standardized name from OpenGraph? In the context of articles, `published` seems more appropriate than `created`. I am guessing that Schema.org is way more generic, so they went with a `dateCreated`.

Adding the `_AT` suffix makes it clear to callers what the format is, and removes the distinction between `Date` (without time) and `DateTime` since ISO 8601 defines both in the same format.

`CREATED_AT` would also be fine, WDYT?

ciferkey replied:
Agree with the naming change. I had arbitrarily chosen between schema.org's and opengraph's convention. I like your reasoning for the names. Will update.

chimbori commented on src/main/kotlin/com/chimbori/crux/api/Fields.kt:
```diff
@@ -8,6 +8,8 @@ public object Fields {
   public const val DISPLAY: String = "display"
   public const val ORIENTATION: String = "orientation"
+  public const val CREATED: String = "created"
+  public const val MODIFIED: String = "modified"
```
`MODIFIED_AT`

chimbori commented on src/main/kotlin/com/chimbori/crux/articles/Article.kt:
```diff
@@ -12,6 +12,8 @@ public data class Article(
   var description: String? = null,
   var siteName: String? = null,
   var themeColor: String? = null,
+  var created: String? = null,
```
Are you explicitly using the previous API? If not, my recommendation would be to leave the `Article` alone, and just use the new API if you need the newly-added fields.

ciferkey replied:
> use the new API if you need the newly-added fields
Following up on this. I wanted to update the golden file tests to work with the new API but it looks like it only uses the old article API. Is there a way to test against Resource in the golden file tests? Or should I not use the integration tests and add unit tests to the HtmlMetadataExtractorTest?

chimbori replied:
Good point, I hadn’t yet gotten around to rewriting those tests using the new API, but now’s a good start!

You can use this test as a sample for your own!

https://github.com/chimbori/crux/commit/07db8c83ca96799051c153a1f81310aadcae824d#diff-c072d40a69e184ed86d22d4fe351ed96c5262bb5525dc8ffa8fc325f19f77edc

chimbori commented on src/main/kotlin/com/chimbori/crux/plugins/HtmlMetadataExtractor.kt:
```diff
@@ -61,6 +65,8 @@ public class HtmlMetadataExtractor(private val okHttpClient: OkHttpClient) : Ext
         DESCRIPTION to resourceToUse.document?.extractDescription(),
         SITE_NAME to resourceToUse.document?.extractSiteName(),
         THEME_COLOR_HEX to resourceToUse.document?.extractThemeColor(),
+        CREATED to resourceToUse.document?.extractCreated(),
```
Should convert to ISO 8601 at this point if not already in ISO 8601 format.

chimbori commented on src/test/kotlin/com/chimbori/crux/articles/GoldenFilesTest.kt:
```diff
@@ -59,6 +59,8 @@ class GoldenFilesTest {
         "A Nigerian woman has given birth to a boy on board a rescue ship in the Mediterranean after being plucked from an overcrowded rubber dinghy.",
         document?.text()
       )
+      assertEquals("2016-09-12T14:31:25+00:00", created)
```
Awesome, thanks for adding tests!

chimbori committed:
Add a sample test for golden files using the new Resource API

Eventually, all GoldenFileTests will be moved to use the new Resource API, and the older Article API will be marked deprecated.

chimbori committed:
Add a sample test for golden files using the new Resource API

Eventually, all GoldenFileTests will be moved to use the new Resource API, and the older Article API will be marked deprecated.

ciferkey committed:
updates following PR feedback:
 - removed NBC example and non standard metadata
 - reverted changes to old article extractor API
 - renamed new fields to PUBLISHED_AT and MODIFIED_AT

ciferkey committed:
updates following PR feedback:
 - removed NBC example and non standard metadata
 - reverted changes to old article extractor API
 - renamed new fields to PUBLISHED_AT and MODIFIED_AT

ciferkey committed:
reverted change to Article.kt

ciferkey committed:
reverted change to Article.kt

ciferkey committed:
Merge remote-tracking branch 'origin/main'

ciferkey commented:
Thank you for adding that example test.

I've rebased on top of your new version and created new tests for the test cases I had previously added to GoldenFileTest.

Let me know if there are any additional changes you would like.

chimbori reviewed (approved):
Looks great, thanks for the PR, and I’m sure a lot of users will be excited about these changes!

chimbori commented:
Just published https://github.com/chimbori/crux/releases/tag/v3.9.0 with your changes.

Maven Central / Sonatype seems to be having issues accepting the update, so I’ll retry again later.

ciferkey commented:
Thanks. Excited to try out the changes.

ciferkey commented:
Re: the issue with the update being accepted. On jitpack the 3.9.0 build failed and it provides the following [error](https://jitpack.io/com/github/chimbori/crux/v3.9.0/build.log):

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileKotlin'.
> Error while evaluating property 'filteredArgumentsMap' of task ':compileKotlin'
   > Could not resolve all files for configuration ':compileClasspath'.
      > Could not resolve com.beust:klaxon:5.6.
        Required by:
            project :
         > No matching variant of com.beust:klaxon:5.6 was found. The consumer was configured to find an API of a library compatible with Java 8, preferably in the form of class files, preferably optimized for standard JVMs, and its dependencies declared externally, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm' but:
             - Variant 'apiElements' capability com.beust:klaxon:5.6 declares an API of a library, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm':
                 - Incompatible because this component declares a component compatible with Java 11 and the consumer needed a component compatible with Java 8
             - Variant 'runtimeElements' capability com.beust:klaxon:5.6 declares a runtime of a library, packaged as a jar, preferably optimized for standard JVMs, and its dependencies declared externally, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm':
                 - Incompatible because this component declares a component compatible with Java 11 and the consumer needed a component compatible with Java 8
```

Dropping it here in case it helps you resolve it.

chimbori commented:
Jitpack is finicky, I don’t usually use or check that.

Please continue to use Maven Central as the canonical & the only officially supported channel.

Thanks!
